lib/clamp-link.in

```diff
@@ -404,19 +404,24 @@ if [ ${#LINK_KERNEL_ARGS[@]} != 0 ]; then
   CLANG_OFFLOAD_BUNDLER_INPUTS_ARGS="-inputs=$TEMP_DIR/__empty.o"
   CLANG_OFFLOAD_BUNDLER_TARGETS_ARGS="-targets=host-x86_64-unknown-linux"
 
+  declare -a pids
   # for each GPU target, lower to GCN ISA in HSACO format
   for AMDGPU_TARGET in ${AMDGPU_TARGET_ARRAY[@]}; do
-    { $CLAMP_DEVICE $TEMP_DIR/kernel.bc $TEMP_DIR/kernel-$AMDGPU_TARGET.hsaco --amdgpu-target=$AMDGPU_TARGET;
+    { $CLAMP_DEVICE $TEMP_DIR/kernel.bc $TEMP_DIR/kernel-$AMDGPU_TARGET.hsaco --amdgpu-target=$AMDGPU_TARGET; } &
 
     # error handling
-    ret=$ret+$?;
+    pids+=("${pids[@]}" "$!")
 
     # augment arguments to clang-offload-bundler
-    CLANG_OFFLOAD_BUNDLER_INPUTS_ARGS+=",$TEMP_DIR/kernel-$AMDGPU_TARGET.hsaco";
-    CLANG_OFFLOAD_BUNDLER_TARGETS_ARGS+=",hcc-amdgcn--amdhsa-$AMDGPU_TARGET"; } &
+    CLANG_OFFLOAD_BUNDLER_INPUTS_ARGS+=",$TEMP_DIR/kernel-$AMDGPU_TARGET.hsaco"
+    CLANG_OFFLOAD_BUNDLER_TARGETS_ARGS+=",hcc-amdgcn--amdhsa-$AMDGPU_TARGET"
   done
 
-  wait
+  # collect all the error codes from forked processes
+  # error handling
+  for pid in ${pids[*]}; do
+    wait $pid || ret=$((ret+$?))
```

Comment:
@aaronenyeshi if i understand this correct, in case `any` target hits an error we'll abort the process, right?

Reply:
Hmm, I don't think it will exit on error, rather it will keep error any exit when it detects it, same as before

Reply:
Maybe I should put the if [$ret !=0] statement inside this loop

Reply:
i think it's better we wait on all subprocess to complete and determine whether to abort.

Reply:
Yes, this implementation will satisfy that goal

Reply:
If a sub-process exited before wait is called, could you make sure that wait returns the correct return status of the terminated process and not 127?

---
man bash:

       wait [-n] [n ...]
              Wait for each specified child process and return its termination status.  Each n may be a process ID or a job specification; if a job spec is given, all processes in that job's pipeline are waited for.  If  n
              is  not  given, all currently active child processes are waited for, and the return status is zero.  If the -n option is supplied, wait waits for any job to terminate and returns its exit status.  If n speci‐
              fies a non-existent process or job, the return status is 127.  Otherwise, the return status is the exit status of the last process or job waited for.
---

Reply:
Right, there is no guarantee that the sub-process won't exit before we call wait. One possible solution would be to put a sleep on the sub-process so that all the sub-processes could be invoked then we will wait on them before they wake up/potentially terminate.

Reply:
having a subprocess to sleep is messy. if we have to sleep i would think it'd be the master process (`clamp-link`) to do that. would this post offer some help?
https://stackoverflow.com/questions/356100/how-to-wait-in-bash-for-several-subprocesses-to-finish-and-return-exit-code-0

Reply:
@scchan Actually i just tested it, and it seems that even if I sleep the main process longer than the sub-processes, the exit code is still maintained until wait is called. This was my test code:
```
  7   ret=0;
  8   declare -a AMDGPU_TARGET_ARRAY=('gfx803' 'gfx900')
  9   declare -a pids
 12  { sleep 3; echo --amdgpu-target=${AMDGPU_TARGET_ARRAY[0]}; exit 3; } &
 13  pids=("${pids[@]}" "$!")
 14  { sleep 1; echo --amdgpu-target=${AMDGPU_TARGET_ARRAY[1]}; exit 4; } &
 16  pids=("${pids[@]}" "$!")
 17  sleep 10;
 22
 23   for pid in ${pids[*]}; do
 24     wait $pid || ret=$((ret+$?))
 25   done
 26
 29   echo "ret: $ret"
 30   echo "pids: ${pids[@]}"
```
Output looks like this:
```
root@468e3189356e:~/hcc# ./test.sh
--amdgpu-target=gfx900
--amdgpu-target=gfx803
ret: 7
pids: 6791 6792
```

Reply:
It seems that the sub-processes that we fork won't fall under `non-existent process or job, the return status is 127`. Seems for sub-processes even when finished before wait will stay active as a `zombie` process until the parent waits on it or is done

Reply:
@whchung After experimentation, it should work without the sleeps